### PR TITLE
add total superfluid delegation sum invariant

### PR DESCRIPTION
**Pull request by antstalepresh**
Closes: #XXX

## Description

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/gaia/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: x/superfluid/keeper/intermediary_account.go [hunk 1]
```diff
@@ -113,11 +113,11 @@ func (k Keeper) DeleteIntermediaryAccount(ctx sdk.Context, address sdk.AccAddres
 	prefixStore.Delete(address)
 }
 
-func (k Keeper) SetLockIdIntermediaryAccountConnection(ctx sdk.Context, lockId uint64, acc types.SuperfluidIntermediaryAccount) {
+func (k Keeper) SetLockIdIntermediaryAccountConnection(ctx sdk.Context, lockId uint64, acc sdk.AccAddress) {
```
Comment (ValarDragon) [inline x/superfluid/keeper/intermediary_account.go:116]: Can we keep it as before, I prefer the type safety it implied. (And instead make the genesis logic do the validation to get the superfluid intermediary account)

File: x/superfluid/keeper/invariants.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	}
+}
+
+// TotalSuperfluidDelegationInvariant checks the sum of intermediary account delegation is same as sum of individual lockup delegation
+func TotalSuperfluidDelegationInvariant(keeper Keeper) sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		accs := keeper.GetAllIntermediaryAccounts(ctx)
+		totalSuperfluidDelegationTokens := sdk.ZeroDec()
+
+		for _, acc := range accs {
```
Comment (ValarDragon) [inline x/superfluid/keeper/invariants.go:29]:
```suggestion
		// Compute the total amount delegated from all intermediary accounts
		for _, acc := range accs {
```

File: x/superfluid/keeper/invariants.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+			delegation, found := keeper.sk.GetDelegation(ctx, acc.GetAccAddress(), valAddr)
+			if found {
+				tokens := validator.TokensFromShares(delegation.Shares)
+				totalSuperfluidDelegationTokens = totalSuperfluidDelegationTokens.Add(tokens)
+			}
+		}
+
+		totalExpectedSuperfluidAmount := sdk.ZeroInt()
+		connections := keeper.GetAllLockIdIntermediaryAccountConnections(ctx)
+		for _, connection := range connections {
```
Comment (ValarDragon) [inline x/superfluid/keeper/invariants.go:49]:
```suggestion
		// Compute the total delegation amount expected
		// from every lockID intermediary account connections
		totalExpectedSuperfluidAmount := sdk.ZeroInt()
		connections := keeper.GetAllLockIdIntermediaryAccountConnections(ctx)
		for _, connection := range connections {
```

File: x/superfluid/keeper/invariants.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+					"\tonly single coin lockup is eligible for superfluid staking"), true
+			}
+			amount := keeper.GetSuperfluidOSMOTokens(ctx, lock.Coins[0].Denom, lock.Coins[0].Amount)
+			totalExpectedSuperfluidAmount = totalExpectedSuperfluidAmount.Add(amount)
+		}
+
+		if !totalExpectedSuperfluidAmount.ToDec().Equal(totalSuperfluidDelegationTokens) {
+			return sdk.FormatInvariant(types.ModuleName, totalSuperfluidDelegationInvariantName,
+				"\ttotal superfluid intermediary account delegation amount does not match total sum of lockup delegations\n"), true
+		}
```
Comment (ValarDragon) [inline x/superfluid/keeper/invariants.go:67]:
```suggestion
		// Why do we not expect there to be any rounding error here?
		if !totalExpectedSuperfluidAmount.ToDec().Equal(totalSuperfluidDelegationTokens) {
			return sdk.FormatInvariant(types.ModuleName, totalSuperfluidDelegationInvariantName,
				"\ttotal superfluid intermediary account delegation amount does not match total sum of lockup delegations\n"), true
		}
```

Comment (ValarDragon) [inline x/superfluid/keeper/invariants.go:67]: Why can we expect no rounding error here?

**Review by ValarDragon — CHANGES_REQUESTED** [review]
LGTM to merge once the API comment is resolved

File: x/superfluid/keeper/invariants.go (hunk 4 above)
Comment (antstalepresh) [reply x/superfluid/keeper/invariants.go]: For this, changed the ToDec to RoundInt here https://github.com/osmosis-labs/osmosis/pull/1018/commits/d7f844e75e5237efe571a707ac756255f896680f

File: go.mod [hunk 5]
```diff
@@ -126,10 +126,10 @@ require (
 )
 
 replace (
-    // branch: v0.22.0-osmo-v7, current tag: v0.22.0-osmo-v7.2.0
+	// branch: v0.22.0-osmo-v7, current tag: v0.22.0-osmo-v7.2.0
 	github.com/CosmWasm/wasmd => github.com/osmosis-labs/wasmd v0.22.0-osmo-v7.2
 	// Our cosmos-sdk branch is:  https://github.com/osmosis-labs/cosmos-sdk v0.45.0x-osmo-v7
-	github.com/cosmos/cosmos-sdk => github.com/osmosis-labs/cosmos-sdk v0.45.1-0.20220221115656-75545ea8cb2d
+	github.com/cosmos/cosmos-sdk => github.com/osmosis-labs/cosmos-sdk v0.45.1-0.20220309174635-d81fb5657a15
```
Comment (ValarDragon) [inline go.mod:132]: Can you PR the hook addition to the SDK branch, then we can get this merged?

Comment (antstalepresh) [reply go.mod:132]: Here's the PR for after slash hook
https://github.com/osmosis-labs/cosmos-sdk/pull/136/files

**Comment by ValarDragon** [comment]
Just needs to update the go.mod reference to be merged commit in the SDK branch, and we can get this merged.

**Comment by antstalepresh** [comment]
Here's cherry picked version of this PR that is to main branch. And closing this.
https://github.com/osmosis-labs/osmosis/pull/1081

**Pull request closed without merging**